src/amo/store.js

```diff
@@ -28,11 +29,62 @@ import redirectTo from 'core/reducers/redirectTo';
 import search from 'core/reducers/search';
 import uiState from 'core/reducers/uiState';
 import { middleware } from 'core/store';
+import type { AddonsByAuthorsState } from 'amo/reducers/addonsByAuthors';
+import type { CollectionsState } from 'amo/reducers/collections';
+import type { HomeState } from 'amo/reducers/home';
+import type { RecommendationsState } from 'amo/reducers/recommendations';
+import type { ReviewsState } from 'amo/reducers/reviews';
+import type { UserAbuseReportsState } from 'amo/reducers/userAbuseReports';
+import type { UsersState } from 'amo/reducers/users';
+import type { ViewContextState } from 'amo/reducers/viewContext';
+import type { AbuseState } from 'core/reducers/abuse';
+import type { AddonsState } from 'core/reducers/addons';
+import type { ApiState } from 'core/reducers/api';
+import type { ErrorPageState } from 'core/reducers/errorPage';
+import type { FormOverlayState } from 'core/reducers/formOverlay';
+import type { LanguageToolsState } from 'core/reducers/languageTools';
+import type { InstallationsState } from 'core/reducers/installations';
+import type { RedirectToState } from 'core/reducers/redirectTo';
+import type { SearchState } from 'core/reducers/search';
+import type { UIStateState } from 'core/reducers/uiState';
+
+export type AppState = {|
```

Comment:
This is super great but if we add a new reducer and forget to update `AppState`, Flow will not show an error.

After some quick digging, I found a way to solve that. It's a little cryptic (!) but I added comments to explain what's happening. I tested removing / adding a reducer and Flow does show an error. Flow also wanted me to make a small change to the `reviews` reducer. 

Can you add this patch?

```diff
diff --git a/.eslintrc b/.eslintrc
index 2687f1a3b..4d09ac1f1 100644
--- a/.eslintrc
+++ b/.eslintrc
@@ -8,6 +8,8 @@
     "document": true,
     "webpackIsomorphicTools": true,
     "ga": true,
+    // https://flow.org/en/docs/types/utilities/#toc-objmap
+    "$ObjMap": true,
     // https://flow.org/en/docs/types/utilities/#toc-propertytype
     "$PropertyType": true,
     // These are undocumented utilities that ship with Flow
diff --git a/src/amo/reducers/reviews.js b/src/amo/reducers/reviews.js
index eaca60906..f5161eba7 100644
--- a/src/amo/reducers/reviews.js
+++ b/src/amo/reducers/reviews.js
@@ -167,7 +167,7 @@ export const changeViewState = ({
   state,
   reviewId,
   stateChange,
-}: ChangeViewStateParams = {}): $Shape<ReviewsState> => {
+}: ChangeViewStateParams = {}): ReviewsState => {
   const change = { ...stateChange };
 
   const existingFlag = state.view[reviewId] ? state.view[reviewId].flag : {};
diff --git a/src/amo/store.js b/src/amo/store.js
index 411554cb0..f7fd2106e 100644
--- a/src/amo/store.js
+++ b/src/amo/store.js
@@ -76,6 +76,18 @@ export type AppState = {|
   viewContext: ViewContextState,
 |};
 
+// This is a type function that takes a state type and returns a reducer
+// type, i.e. a function that accepts and returns the same state type.
+/* eslint-disable no-undef */
+type CreateReducerType = <AnyState>(
+  AnyState,
+) => (AnyState, action: Object) => AnyState;
+/* eslint-enable no-undef */
+
+// Given AppState, create a type for all possible application reducers.
+// See https://flow.org/en/docs/types/utilities/#toc-objmap
+type AppReducersType = $ObjMap<AppState, CreateReducerType>;
+
 type CreateStoreParams = {|
   history: Object,
   initialState: Object,
@@ -86,34 +98,35 @@ export default function createStore({
   initialState = {},
 }: CreateStoreParams = {}) {
   const sagaMiddleware = createSagaMiddleware();
+  const reducers: AppReducersType = {
+    abuse,
+    addons,
+    addonsByAuthors,
+    api,
+    autocomplete,
+    categories,
+    collections,
+    errors,
+    errorPage,
+    formOverlay,
+    heroBanners,
+    home,
+    infoDialog,
+    installations,
+    landing,
+    languageTools,
+    recommendations,
+    redirectTo,
+    reviews,
+    routing,
+    search,
+    uiState,
+    userAbuseReports,
+    users,
+    viewContext,
+  };
   const store = _createStore(
-    combineReducers({
-      abuse,
-      addons,
-      addonsByAuthors,
-      api,
-      autocomplete,
-      categories,
-      collections,
-      errors,
-      errorPage,
-      formOverlay,
-      heroBanners,
-      home,
-      infoDialog,
-      installations,
-      landing,
-      languageTools,
-      recommendations,
-      redirectTo,
-      reviews,
-      routing,
-      search,
-      uiState,
-      userAbuseReports,
-      users,
-      viewContext,
-    }),
+    combineReducers(reducers),
     initialState,
     middleware({
       routerMiddleware: routerMiddleware(history),

```

Reply:
ah good call, I used the new type on this object too but then flow complained too much so I removed it...